File: Realize/6Week/PracticeTree3.cpp
#include <iostream>
#include <queue>
#include <vector>
using namespace std;

class Node {
public:
    char data;
    Node* left;
    Node* right;
    Node* parent;

    Node(char data) {
        this->data = data;
        this->left = nullptr;
        this->right = nullptr;
        this->parent = nullptr;
    }
};

class BinaryTree {
public:
    Node* head;

    BinaryTree(vector<char>& arr) {
        vector<Node*> nodes(arr.size());
        for (int i = 0; i < arr.size(); i++) {
            nodes[i] = new Node(arr[i]);
        }

        for (int i = 0; i < arr.size(); i++) {
            int left = 2 * i + 1;
            int right = 2 * i + 2;

            if (left < arr.size()) {
                nodes[i]->left = nodes[left];
                nodes[left]->parent = nodes[i];
            }
            if (right < arr.size()) {
                nodes[i]->right = nodes[right];
                nodes[right]->parent = nodes[i];
            }
        }
        head = nodes[0];
    }

    Node* searchNode(char data) {
        queue<Node*> q;
        q.push(head);
        while (!q.empty()) {
            Node* cur = q.front();
            q.pop();
            if (cur->data == data) {
                return cur;
            }
            if (cur->left) q.push(cur->left);
            if (cur->right) q.push(cur->right);
        }
        return nullptr;
    }

    int checkSize(char data) {
        Node* node = searchNode(data);
        if (!node) return 0;

        queue<Node*> q;
        q.push(node);
        int size = 0;

        while (!q.empty()) {
            Node* cur = q.front();
            q.pop();
            if (cur->left) {
                q.push(cur->left);
                size++;
            }
            if (cur->right) {
                q.push(cur->right);
                size++;
            }
        }
        return size + 1;
    }
};

int main() {
    vector<char> arr(10);
    for (int i = 0; i < 10; i++) {
        arr[i] = 'A' + i;
    }

    BinaryTree bt(arr);

    cout << "Root: " << bt.head->data << endl;

    Node* B = bt.searchNode('B');
    if (B->left) cout << "B -> left child: " << B->left->data << endl;
    if (B->right) cout << "B -> right child: " << B->right->data << endl;

    Node* F = bt.searchNode('F');
    cout << "F -> parent: " << F->parent->data << endl;

    cout << "Leaf node: ";
    for (char c = 'A'; c <= 'J'; c++) {
        Node* cur = bt.searchNode(c);
        if (cur && !cur->left && !cur->right) {
            cout << cur->data << " ";
        }
    }
    cout << endl;

    Node* E = bt.searchNode('E');
    Node* cur = E;
    int cnt = 0;
    while (cur->parent) {
        cur = cur->parent;
        cnt++;
    }
    cout << "E depth: " << cnt << endl;

    cout << "Level2 Node: ";
    for (char c = 'A'; c <= 'J'; c++) {
        Node* target = bt.searchNode(c);
        cur = target;
        cnt = 0;
        while (cur && cur->parent) {
            cur = cur->parent;
            cnt++;
        }
        if (cnt == 2) {
            cout << target->data << " ";
        }
    }
    cout << endl;

    int maxDepth = 0;
    for (char c = 'A'; c <= 'J'; c++) {
        Node* target = bt.searchNode(c);
        cur = target;
        cnt = 0;
        while (cur && cur->parent) {
            cur = cur->parent;
            cnt++;
        }
        maxDepth = max(maxDepth, cnt);
    }
    cout << "Height: " << maxDepth << endl;

    int size = bt.checkSize('B');
    cout << "B size = " << size << endl;

    return 0;
}

File: Realize/6Week/PracticeTree3.py
from collections import deque

class Node:
    def __init__(self, data):
        self.data = data
        self.left = None
        self.right = None
        self.parent = None

class BinaryTree:
    def __init__(self, arr):
        nodes = [Node(data) for data in arr]

        for i in range(len(arr)):
            left = 2 * i + 1
            right = 2 * i + 2
            if left < len(arr):
                nodes[i].left = nodes[left]
                nodes[left].parent = nodes[i]
            if right < len(arr):
                nodes[i].right = nodes[right]
                nodes[right].parent = nodes[i]

        self.head = nodes[0]

    def search_node(self, data):
        queue = deque([self.head])
        while queue:
            cur = queue.popleft()
            if cur.data == data:
                return cur
            if cur.left:
                queue.append(cur.left)
            if cur.right:
                queue.append(cur.right)
        return None

    def check_size(self, data):
        node = self.search_node(data)
        if not node:
            return 0

        queue = deque([node])
        size = 0

        while queue:
            cur = queue.popleft()
            if cur.left:
                queue.append(cur.left)
                size += 1
            if cur.right:
                queue.append(cur.right)
                size += 1
        return size + 1

if __name__ == "__main__":
    arr = [chr(ord('A') + i) for i in range(10)]
    bt = BinaryTree(arr)

    print(f"Root: {bt.head.data}")

    B = bt.search_node('B')
    if B.left:
        print(f"B -> left child: {B.left.data}")
    if B.right:
        print(f"B -> right child: {B.right.data}")

    F = bt.search_node('F')
    print(f"F -> parent: {F.parent.data}")

    print("Leaf node: ", end="")
    for c in arr:
        cur = bt.search_node(c)
        if cur.left is None and cur.right is None:
            print(cur.data, end=" ")
    print()

    E = bt.search_node('E')
    cur = E
    cnt = 0
    while cur.parent:
        cur = cur.parent
        cnt += 1
    print(f"E depth: {cnt}")

    print("Level2 Node: ", end="")
    for c in arr:
        target = bt.search_node(c)
        cur = target
        cnt = 0
        while cur.parent:
            cur = cur.parent
            cnt += 1
        if cnt == 2:
            print(target.data, end=" ")
    print()

    max_depth = 0
    for c in arr:
        target = bt.search_node(c)
        cur = target
        cnt = 0
        while cur.parent:
            cur = cur.parent
            cnt += 1
        max_depth = max(max_depth, cnt)
    print(f"Height: {max_depth}")

    size = bt.check_size('B')
    print(f"B size = {size}")
